pkgs/by-name/si/sile/package.nix

```diff
@@ -115,6 +115,9 @@ stdenv.mkDerivation (finalAttrs: {
       ps:
       with ps;
       [
+        # used for module detection, also recommended at runtime for 3rd party module installation
+        luarocks
```

Comment:
Hmm I'm not sure about this... The idea is to enable the user to add additional lua modules? If yes, they I'd recommend a completely new approach, that fits the Nix culture:

Create 2 packages: `sile` and `sile.passthru.unwrapped`. The later won't be functional outside a `LUA_{,C}PATH` environment that includes at least the packages listed below. The former will wrap the `sile` binary of the later, such that the top-level attribute will be functional out of the box ("batteries included"), and the later will be used by advanced users who know how to define their environment such that it will include the lua dependencies etc. (see https://direnv.net/ for example).

What do you think of the idea?

Reply:
That *is* something we could setup because yes, users should be able to add modules to their environment. Typically this would be per-project so it isn't a big deal to not have this setup in the Nix package, but it would certainly be more idiomatic for Nix users for it to work that way.

However that was *not* what I was trying to accomplish here though. This is just to get the build working since `luarocks` became a compile time dependency by default in this release. It is used by the `./configure` script to check whether all the default dependencies are installed to the system properly. The configure phase could be configured to work around that task, but it was easier to just let it have it.

Reply:
If `luarocks` is only a build dependency, I'd put it in `nativeBuildInputs`. So it won't be propagated.

As for the wrapping etc, I just noticed that the build doesn't use Nix's wrapping mechanisms, but rather the build itself uses the Lua environment /interpreter that was found during build time to compile documents. Not only that, I think I observed correctly that if environment variables `LUA_{,C}PATH` are set to something, the `sile` executable will use them? If I observed that correctly, these are very good news, and it means that Sile is naturally supporting the declarative environment approach, and that we don't need to reform the expression into a wrapped and unwrapped derivation.

While trying to verify that putting `luarocks` in `nativeBuildInputs` works as expected, I noticed a slightly non-trivial detail there, and I also fixed another minor detail. So my alternative PR is at #359859. Please confirm the above observation, and perhaps we can even slightly modify the attributes in the `passthru` to allow advanced users to use a different Lua environment when using Sile while most importantly, it won't require them to recompile `sile` just due to an additional Lua package.

Reply:
> If `luarocks` is only a build dependency, I'd put it in `nativeBuildInputs`. So it won't be propagated.

Yes! That's better indeed.

> As for the wrapping etc, I just noticed that the build doesn't use Nix's wrapping mechanisms, but rather the build itself uses the Lua environment /interpreter that was found during build time to compile documents.

Yes, and a lot of trouble it was to get that working! Vendoring some runtime of our own would have been easy, but then it would be a never ending fiasco because it would never match the host system. Using the host system's Lua sources to embed a VM and keeping the system default module loading paths with us for later use makes it way more versatile to *use* on most systems.

> Not only that, I think I observed correctly that if environment variables `LUA_{,C}PATH` are set to something, the `sile` executable will use them?

Correct.

> If I observed that correctly, these are very good news, and it means that Sile is naturally supporting the declarative environment approach, and that we don't need to reform the expression into a wrapped and unwrapped derivation.

:+1:

> While trying to verify that putting `luarocks` in `nativeBuildInputs` works as expected, I noticed a slightly non-trivial detail there, and I also fixed another minor detail. So my alternative PR is at #359859.

LGTM, I'll close this one.

> Please confirm the above observation, and perhaps we can even slightly modify the attributes in the `passthru` to allow advanced users to use a different Lua environment when using Sile while most importantly, it won't require them to recompile `sile` just due to an additional Lua package.

That would be fantastic.

Also if possible we should probably set the *default* Lua for the derivation to be LuaJIT. We discussed doing that before the Rust makover, but some of the above bits mentioned details were not fully worked out then.

If it can't be the default for some reason I should at least better document upstream how to override the default and use it that way since it is a *huge* speed boost in production.

Reply:
There was no real issue with using `luajit` by default. It is now implemented. The only slight drawback is that you have to edit the too large `all-packages.nix` file, a file we try to avoid editing as much as possible.

I didn't change the `lua` argument in the top of the expression to `luajit` to keep the current function arguments the same.